#2502 - Add jinja2 template handling to request body

cdubz opened the pull request:
Closes #2502

👋 This is an attempt to add jinja2 handling to request body (and headers). So far I've only got body here and want to confirm the implementation looks reasonable. Headers appears a bit more complicated, but should be doable.

I could also use some advice on what kind of tests to add. For jinja in the URL, I don't see anything in the requests tests. There is some testing in the jinja tests. Any opinion on where I should add a test?

dgtlmoon commented:
nice one

So add your little bit of Jinja2 to this (`Test Body Value`)

https://github.com/dgtlmoon/changedetection.io/blob/e84de7e8f4bb14fac4f8b313e3e7ffdf66981877/changedetectionio/tests/test_request.py#L120C19-L120C33

then rewrite this to be the actual expected jinaj2 body rendered text

https://github.com/dgtlmoon/changedetection.io/blob/e84de7e8f4bb14fac4f8b313e3e7ffdf66981877/changedetectionio/tests/test_request.py#L143-L144

dgtlmoon commented:
Note, other places in that same test its either looking for the original string, or your jinja2 rendered string, so you'll have to step through it

dgtlmoon commented on changedetectionio/forms.py:
```diff
@@ -536,6 +536,20 @@ def validate(self, **kwargs):
             result = False
 
+        # Attempt to validate jinja2 templates in the body
+        if self.body.data:
```
```suggestion
        if self.body.data and self.body.data.strip():
```

dgtlmoon commented:
```
        if request_body:
            from changedetectionio.safe_jinja import render as jinja_render
            request_body = jinja_render(template_str=self.watch.get('body'))
```
Maybe just..

```
  from changedetectionio.safe_jinja import render as jinja_render
  request_body = jinja_render(template_str=self.watch.get('body',''))
```

?

cdubz commented:
> ```
>         if request_body:
>             from changedetectionio.safe_jinja import render as jinja_render
>             request_body = jinja_render(template_str=self.watch.get('body'))
> ```
> 
> Maybe just..
> 
> ```
>   from changedetectionio.safe_jinja import render as jinja_render
>   request_body = jinja_render(template_str=self.watch.get('body',''))
> ```
> 
> ?

I was wondering about that, but wasn't sure if `request_body` being an empty string would impact something else along the way? If not, yeah that seems a little nicer. Maybe it already is, I didn't check 😁

dgtlmoon commented:
OK :) Will wait for you to change those tests and then i'll look into it more

cdubz commented:
Changes pushed. If all that passes, I'll take a look at headers next.

dgtlmoon commented:
you can run it locally

```
cd changedetectionio
pytest ./tests/test_requests.py
```

dgtlmoon commented on changedetectionio/processors/__init__.py:
```diff
@@ -120,7 +120,8 @@ def call_browser(self, preferred_proxy_id=None):
 
-        request_body = self.watch.get('body')
+        from changedetectionio.safe_jinja import render as jinja_render
+        request_body = request_body = jinja_render(template_str=self.watch.get('body', ''))
```
`request_body = request_body` ?

cdubz replied:
🤦

dgtlmoon commented on changedetectionio/tests/test_request.py:
```diff
@@ -140,8 +141,9 @@ def test_body_in_request(client, live_server, measure_memory_usage):
-    # We should see what we sent in the reply
-    assert str.encode(body_value) in res.data
+    # We should see the formatted value of what we sent in the reply
+    assert str.encode(body_value) not in res.data
```
really nice

cdubz commented:
OK, should be good now. I was sneaking this in before work work and didn't run the test on the last change 😁 

Turns out `jinja_render(template_str=self.watch.get('body', ''))` does actually cause test failures so I switched back to the conditional.

cdubz commented:
> Can you add the headers changes to this branch also?

Yep -- thats the plan. Will come back to this after my work day 😁

cdubz commented:
Ok -- this is review ready now with headers support as well!

I also made some tweaks to the edit form to indicate support. Can rollback or adjust those per preference. https://github.com/dgtlmoon/changedetection.io/wiki/Handling-variables-in-the-watched-URL should be updated to indicate support in more than just the URL if they stick.

dgtlmoon commented:
thanks again, theres some false error in the test, i'll keep pushing

cdubz commented:
Oh one thing I just realized I missed — validation of jinja2 for the headers. Want me to add that?

dgtlmoon commented:
yes please! always validate, make sure bad jinja2 validation throws an exception that can be seen from the watch overview list

cdubz commented:
Added!

I also added a separate commit to improve the error messages on the jinja validation from these fields. The results look like this:

<img width="675" alt="image" src="https://github.com/user-attachments/assets/9b69bcb4-1bfa-482d-9860-1805029be921">

The change also covers the URL field.

I didn't see this being done anywhere else, so easy to reverse if it's not wanted.

dgtlmoon commented:
really nice!!

cdubz commented on changedetectionio/forms.py:
```diff
@@ -529,7 +529,7 @@ def validate(self, **kwargs):
             # incase jinja2_time or others is missing
             logger.error(e)
-            self.headers.errors.append(f'Invalid template syntax configuration: {e}')
+            self.url.errors.append(f'Invalid template syntax configuration: {e}')
```
I wondered about that too since I didn't see a pattern like this used elsewhere. It appears the underlying classes protect against this. E.g. --

<img width="681" alt="image" src="https://github.com/user-attachments/assets/ac6da75e-eb24-437f-a933-bcc2baf7c8a6">

<img width="863" alt="image" src="https://github.com/user-attachments/assets/7884ae73-b818-4b61-8d0f-fd86aa4e02bc">

dgtlmoon commented:
amazing thank you so much <3
